### ISSUE TITLE
Add Inference support to PyTorch Lightning

### ISSUE THREAD

**Issue by tchaton** [Lightning-AI/lightning#13480]
## 🚀 Feature

From my current investigation, it seems TorchServe has matured and it is now the default Server used within Sagemaker AWS PyTorch Image.

Until now, PyTorch Lightning has offered very limited support for serving and the transition from training to production has been quite challenging for many users.

This issue is meant to start a conversation on how PyTorch Lightning can smooth the transition between those worlds.  

My proposition is the following:

Add a new hook to the LightningModule called `configure_serve_handler` where users can return a [BaseHandler](https://pytorch.org/serve/_modules/ts/torch_handler/base_handler.html#BaseHandler) subclass to accommodate their models' serving requirements.

Here is how it could look like:

```py
class MNISTClassifier(LightningModule):

    def configure_serve_handler(self):
        return MNISTDigitClassifierHandler(...) 
```

Here are some opportunities available with such an introduction:

* Sanity serve checking when starting training. 
* Sanity TorchScripting when starting training.
* Auto-generate an handler.py file for the users without the pytorch lightning dependency.
* Auto-generate mar file with torchscripted model, handlers, ....
* Provide DataType to help users to write down serialization and deserialization layers for their models or upstream to Serve.

Furthermore, using the newly introduced Lightning App, it would be possible to provide even more to PyTorch Lightning users by automating all those steps and then serving their model to any services of their choice.  

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @tchaton @justusschock @awaelchli @carmocca @ananthsub @ninginthecloud @jjenniferdai @rohitgr7

**Comment by tchaton** [comment]
@justusschock I think it boils down to being creative with the right design.

Another one would be to introduce a serve_step and provides deserialization / serialization data type for the users to annotate the method.

```py
class MNISTClassifier(LightningModule):

    def serve_step(self, x: Image) -> Logits:
        return self.model(y)
```

**Comment by justusschock** [comment]
If we introduce datatypes to PL we are again making assumptions on the data. Currently PL is data agnostic and imo it should stay that way. Otherwise you will be adding more and more datatypes. IMO that’s the domain of flash :)

**Comment by tchaton** [comment]
The way data is encoded in requests is standardized and we aren't making that many assumptions. We are actually removing boilerplate for the users who would have to implement it over and over.
I don't think it is making assumptions especially if the user can easily provide their own. 

Another possible API would be but I think it introduces more boilerplate.

```py
class Model(MyPlModule, ServeModule):

    input = ...
    output = ...

    def payload_exaple():
        return ...

    def serve_step(self, x, y)
        return self. ...

pl_module = Model()

server = Server(pl_module)
server.sanity_check()

trainer = Trainer()
trainer.fit(pl_module)
```

**Comment by carmocca** [comment]
I agree with @justusschock in that going into seralization of data types etc is a big can of worms that should stay out of PL.

To me, the key here is evaluating whether there's a direct interaction between the `Trainer` and serving. The top post proposes sanity checking steps which could be useful in specific cases  but are very niche. How often does serving performance inform the training procedure sort of like a validation mechanism? If it's just for performance tracking, then this can be done in parallel to the training procedure by evaluating checkpoints along the way.

My key concern is, what do we gain from an API like:
```python
class MyLitModule(pl.LightningModule):
    def serve_step():
        # fill in!
        ...
```
vs
```python
pl_module = MyLitModule()
# maybe train with the Trainer...
# do whatever you want

# where does this come from? pl? flash? new project?
server = LightningServe(pl_module)  # you can subclass and override. this contains all the serving magic, datatypes, etc...
server.start()
```

There can be further iterations to this, like a mixin that would tie together the PL module with the serving. But this still keeps it out of the `LightningModule` interface which atm is meant to act as just a contract with the Trainer.

What I think is very important is that we ease the transition OUT of a trained PL module into nn.Modules which is what most of the ecosystem projects understand.

(note: these are my thoughts after discussing with Thomas offline for a while :+1:)

**Comment by tchaton** [comment]
@ethanwharris Would you mind sharing your thoughts on this open conversation?

**Comment by zippeurfou** [comment]
Quick thought on this.
I agree with @tchaton to says that:
1. There is a lot of boilerplate that can be removed on the serving side and we should figure a way with Lightning at large (including Lightning Apps / Component) to make it easier. 
2. Pytorch Lightning should allow the transition to serving to be easy. Assuming you train a model and that's it is awesome in term of focus but doing model training is only one part of the system. The same way we introduced logger, we know this is helpful.

That being said, I agree with @justusschock and @carmocca  to say we have to be careful in terms of what we introduce here to make sure we don't put ourselves in a situation that makes too many assumptions and is not actually realistic.

I would add that one thing PL bring is also the normalization of the code which not only removes boilerplate but also the code to be easier to read and easier to collaborate on.

I do wonder if we can introduce light touch that would allow the same effect for serving while allowing the flexibility needed for the users and not forcing data type. We can then slowly figure out what boilerplate we can remove.

**Comment by lantiga** [comment]
+1 on providing a solution to the issue within Lightning @tchaton.

I personally like @carmocca’s proposal: let’s start from a mixin with its own set of hooks. This way we can develop a clear contract for getting a Lightning model to serving.

Note that we are *not* talking about implementing a serving solution in this context: the proposal here is to have all the hooks needed to do sanity checking, generating a serving script, or export to an existing serving system - TorchServe in the original proposal but it will be pluggable moving forward.

It is about a clear contract on writing the pieces of info needed to get to serving (including eventually data transformations) rather than serving per se. We may still provide out own Server for development purposes (or beyond) but that’s not the centerpiece.

**Comment by lantiga** [comment]
A few extra thoughts:
- transformations of incoming data (normalization etc) are typically quite different during serving compared to training or evaluation
- decoupling serve from the Lightning module may be beneficial since training and serving are often not 1 to 1: it is not uncommon to have multiple models in a single serve pass, we may have a chance to have an expressive API for this
- future serving of PyTorch models from Python will be *fast*, we may be able to really leverage our API for managing optimizations etc once dynamo lands (https://github.com/pytorch/torchdynamo)
